File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
             int left=0, right=0;
       HashMap<Character, Integer>mp=new HashMap<Character, Integer>();
        int len=0;
        int n= s.length();
        while(right<n){
            if(mp.containsKey(s.charAt(right))){
                 left= Math.max(left, mp.get(s.charAt(right))+1);
            }
            
            // update occurence of char 
             mp.put((s.charAt(right)), right);
            len = Math.max(len, right-left+1);
            right++;
        }
        return len;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxAns=0;
        map<char, int>mp;
        int i=0, j=0;
        while(j<s.length()){
            // if present in map then shrink 
            while(i<=j && (mp.find(s[j])!=mp.end())){
                mp[s[i]]--;
                if(mp[s[i]]==0){
                    mp.erase(s[i]);
                }
                i++;
            }
            mp[s[j]]++;
            maxAns= max(maxAns, (int)mp.size());
            j++;
        }
        return maxAns;
    }
};